### PR TITLE
Move expand/expandAs code to backends

### PR DESCRIPTION
**Pull request by killeent**
As title. Addresses #1312. The implementation matches similar methods such as `view`/`viewAs`, which internally calls a function `newView` that returns a newly allocated Tensor backed by the same storage.

Test Plan: Run unit tests.

### CODE REVIEW

**Comment by ngimel** [comment]
I think it would be better to have a helper function calculating expanded sizes and strides in TH, that is called from both TH and THC (similar to view, https://github.com/pytorch/pytorch/blob/master/torch/lib/THC/generic/THCTensor.c#L231), rather than have 2 almost verbatim similar TH(C)Tensor_(newExpand) doing that in TH and THC.

**Comment by gchanan** [comment]
@ngimel good suggestion.  One other point that @apazke made previously is that this can be done more efficiently in a single loop.  I'm implementing something like this for two operands to support efficient numpy style broadcast (see https://github.com/gchanan/pytorch/commit/63711ee75c0f344443fcc54553a9016255e6bd5c as a POC).  Since we probably want to share these implementations as much as possible, and it's not clear exactly what that means yet with two operands, I'd suggest we either just get this in and I'll do the work to unify everything or @kileeent and I work this out in a separate branch.  I'd have a preference for the former given broadcast could take some work (need to decide for each function whether it's supported, update documentation, etc), but I'm happy either way.

File: torch/tensor.py [hunk 1]
```diff
@@ -205,65 +205,6 @@ def permute(self, *dims):
                 perm[j] = -1
         return tensor
 
-    def expand(self, *sizes):
-        """Returns a new view of the tensor with singleton dimensions expanded
```
Comment (colesbury) [inline torch/tensor.py]: Can you move the docs to torch/_torch_docs.py:

```
add_docstr(torch._C.expand, ...)
```

Comment (killeent) [inline torch/tensor.py]: Will do.

**Comment by killeent** [comment]
Per discussion offline with @gchanan, to unblock him on broadcast semantics, I will go ahead and fix the duplicate code, but leave any other optimizations out. He will implement these in a subsequent diff.

**Review by apaszke — APPROVED** [review]

**Pull request merged**